packages/enzyme/src/ShallowWrapper.js

```diff
@@ -88,6 +94,48 @@ function validateOptions(options) {
   ) {
     throw new Error('lifecycleExperimental and disableLifecycleMethods cannot be set to the same value');
   }
+
+  if (
+    typeof enableComponentDidUpdateOnSetState !== 'undefined' &&
+    lifecycles.componentDidUpdate &&
+    lifecycles.componentDidUpdate.onSetState !== enableComponentDidUpdateOnSetState
+  ) {
+    throw new TypeError('the legacy enableComponentDidUpdateOnSetState option should be replaced by `lifecycles: { componentDidUpdate: { onSetState: true } }`');
+  }
+
+  if (
+    typeof supportPrevContextArgumentOfComponentDidUpdate !== 'undefined' &&
+    lifecycles.componentDidUpdate &&
+    lifecycles.componentDidUpdate.prevContext !== supportPrevContextArgumentOfComponentDidUpdate
+  ) {
+    throw new TypeError('the legacy supportPrevContextArgumentOfComponentDidUpdate option should be replaced by `lifecycles: { componentDidUpdate: { prevContext: true } }`');
+  }
+}
+
+function getAdapterLifecycles({ options }) {
+  const {
+    lifecycles = {},
+    enableComponentDidUpdateOnSetState,
+    supportPrevContextArgumentOfComponentDidUpdate,
+  } = options;
+
+  const hasLegacySetStateArg = typeof enableComponentDidUpdateOnSetState !== 'undefined';
+  const hasLegacyPrevContextArg = typeof supportPrevContextArgumentOfComponentDidUpdate !== 'undefined';
+  const componentDidUpdate = hasLegacySetStateArg || hasLegacyPrevContextArg
+    ? {
+      ...(hasLegacySetStateArg && {
+        onSetState: !!enableComponentDidUpdateOnSetState,
+      }),
+      ...(hasLegacyPrevContextArg && {
+        prevContext: !!supportPrevContextArgumentOfComponentDidUpdate,
+      }),
+    }
+    : null;
+
+  return {
+    ...lifecycles,
+    ...(componentDidUpdate && { componentDidUpdate }),
```

Comment:
I could foresee a problem where there is a mismatch between 
```
supportPrevContextArgumentOfComponentDidUpdate: true
``` 
and 
```
lifecycles: {
  ...lifecycles,
  componentDidUpdate: {
    prevContext: false,
  },
},
```

this code above would mean that the legacy settings would the one that is honored. Would we instead want to honor the new lifecycles settings object over the legacy settings if both are present? This isn't a block, but more of a discussion. Also not sure if there is a clean way to do this...

Reply:
I suppose that this would be caught here:

```
typeof enableComponentDidUpdateOnSetState !== 'undefined' &&
lifecycles.componentDidUpdate &&
lifecycles.componentDidUpdate.onSetState !== enableComponentDidUpdateOnSetState
```
but then that makes the error message potentially misleading, as the new setting format is set, but is different.

Reply:
Good call; i'll improve the error message.